src/imitation/scripts/train_adversarial.py

```diff
@@ -114,116 +134,37 @@ def train_and_plot(_seed: int,
-        plt.xlabel("epochs")
-        plt.ylabel("Average reward per episode (n={})"
-                   .format(n_episodes_per_reward_data))
-        plt.plot(gen_ep_reward[name], label="avg gen ep reward", c="red")
-        plt.plot(rand_ep_reward[name],
-                 label="avg random ep reward", c="black")
-        plt.plot(exp_ep_reward[name], label="avg exp ep reward", c="blue")
-        plt.legend()
-        _savefig_timestamp("plot_fight_epreward_gen", interactive)
-
-    if n_epochs_per_plot is not None:
-      n_plots_per_epoch = 1 / n_epochs_per_plot
+    if plot_interval > 0:
+      visualizer = _TrainVisualizer(
+        trainer=trainer,
+        show_plots=show_plots,
+        n_episodes_per_reward_data=n_plot_episodes,
+        log_dir=log_dir,
+        expert_policy=expert_policy_plot)
     else:
-      n_plots_per_epoch = None
-
-    def should_plot_now(epoch) -> bool:
-      """For positive epochs, returns True if a plot should be rendered now.
-
-      This also controls the frequency at which `ep_reward_plot_add_data` is
-      called, because generating those rollouts is too expensive to perform
-      every timestep.
-      """
-      assert epoch >= 1
-      if n_plots_per_epoch is None:
-        return False
-      plot_num = math.floor(n_plots_per_epoch * epoch)
-      prev_plot_num = math.floor(n_plots_per_epoch * (epoch - 1))
-      assert abs(plot_num - prev_plot_num) <= 1
-      return plot_num != prev_plot_num
-
-    # Collect data for epoch 0.
-    if n_epochs_per_plot is not None:
-      disc_plot_add_data(False)
-      ep_reward_plot_add_data(trainer.env, "Ground Truth Reward")
-      ep_reward_plot_add_data(trainer.env_train, "Train Reward")
-      ep_reward_plot_add_data(trainer.env_test, "Test Reward")
+      visualizer = None
 
     # Main training loop.
     for epoch in tqdm.tqdm(range(1, n_epochs+1), desc="epoch"):
       trainer.train_disc(n_disc_steps_per_epoch)
-      disc_plot_add_data(False)
+      if visualizer:
+        visualizer.disc_plot_add_data(False)
+
       trainer.train_gen(n_gen_steps_per_epoch)
-      disc_plot_add_data(True)
+      if visualizer:
+        visualizer.disc_plot_add_data(True)
 
-      if should_plot_now(epoch):
-        disc_plot_show()
-        ep_reward_plot_add_data(trainer.env, "Ground Truth Reward")
-        ep_reward_plot_add_data(trainer.env_train, "Train Reward")
-        ep_reward_plot_add_data(trainer.env_test, "Test Reward")
-        ep_reward_plot_show()
+      if visualizer and epoch % plot_interval == 0:
+        visualizer.disc_plot_show()
+        visualizer.ep_reward_plot_add_data(trainer.env, "Ground Truth Reward")
+        visualizer.ep_reward_plot_add_data(trainer.env_train, "Train Reward")
+        visualizer.ep_reward_plot_add_data(trainer.env_test, "Test Reward")
+        visualizer.ep_reward_plot_show()
+
+      if ray_tune_interval > 0 and epoch % ray_tune_interval == 0:
+        gen_ret = util.rollout.mean_return(
```

Comment:
Potential issues with `VecNormalize` (along with potentially other uses in the code, e.g. `visualizer` logging.)

Reply:
Will try to address this in #90 along with `ray.tune` episode reward mean reports.

Reply:
Nevermind, I will address `VecNormalize` in a separate PR.